adminSiteServer/apiRoutes/dataInsights.ts

```diff
@@ -0,0 +1,294 @@
+        })
+    )
+}
+
+function extractDataInsightIndexItem(
+    dataInsight: DataInsightRow
+): OwidGdocDataInsightIndexItem {
+    const content = parsePostGdocContent(
+        dataInsight.content
+    ) as OwidGdocDataInsightContent
+    const chartConfig = dataInsight.chartConfig
+        ? parseChartConfig(dataInsight.chartConfig)
+        : undefined
+
+    // check if the given grapher-url is a valid Grapher or Explorer URL
+    const grapherUrlField = content["grapher-url"]?.trim()
+    const grapherUrl = grapherUrlField?.startsWith(GRAPHER_URL_PREFIX)
+        ? grapherUrlField
+        : undefined
+    const explorerUrl = grapherUrlField?.startsWith(EXPLORER_URL_PREFIX)
+        ? grapherUrlField
+        : undefined
+
+    // collect narrative chart data if it exists
+    const narrativeChart =
+        content["narrative-chart"] && dataInsight.narrativeChartId
+            ? {
+                  id: dataInsight.narrativeChartId,
+                  name: content["narrative-chart"],
+                  chartConfigId: dataInsight.narrativeChartConfigId!,
+              }
+            : undefined
+
+    // detect the chart type from the narrative chart or grapher URL
+    const chartType = detectChartType({
+        narrativeChart: narrativeChart?.name,
+        grapherUrl,
+        chartConfig,
+    })
+
+    // collect the image data if it exists
+    const image = dataInsight.imageId
+        ? {
+              id: dataInsight.imageId,
+              filename: dataInsight.filename,
+              cloudflareId: dataInsight.cloudflareId,
+              originalWidth: dataInsight.originalWidth,
+          }
+        : undefined
+
+    return {
+        id: dataInsight.gdocId,
+        slug: dataInsight.slug,
+        tags: dataInsight.tags ?? [],
+        published: !!dataInsight.published,
+        publishedAt: dataInsight.publishedAt,
+        markdown: dataInsight.markdown,
+        title: content.title ?? "",
+        authors: content.authors,
+        approvedBy: content["approved-by"],
+        narrativeChart,
+        grapherUrl,
+        explorerUrl,
+        figmaUrl: content["figma-url"],
+        chartType,
+        image,
+    }
+}
+
+function detectChartType({
+    narrativeChart,
+    grapherUrl,
+    chartConfig,
+}: {
+    narrativeChart?: string
+    grapherUrl?: string
+    chartConfig?: GrapherInterface
+}): GrapherChartOrMapType | undefined {
+    if (!chartConfig) return undefined
+
+    if (narrativeChart) return getChartTypeFromConfig(chartConfig)
+
+    if (grapherUrl) {
+        const url = new URL(grapherUrl)
+        return getChartTypeFromConfigAndQueryParams(
+            chartConfig,
+            url.searchParams
+        )
+    }
+
+    return undefined
+}
+
+function getChartTypeFromConfig(
+    chartConfig: GrapherInterface
+): GrapherChartOrMapType | undefined {
+    return getChartTypeFromConfigAndQueryParams(chartConfig)
+}
+
+function getChartTypeFromConfigAndQueryParams(
```

Comment:
Is this the right folder for this function? I'm surprised we don't have something like this already, or that it wouldn't be better to colocate it with other config utils.

In any case, it seems a bit easy to forget to update if we add a new chart type - do you think that's likely to be a problem? And if so, would `ts-pattern`'s  `P.match().exhaustive()` help?

Reply:
Also, oh no, just realizing all of this is for the column I suggested you drop 💀

Reply:
I don't think it's that common, to be honest, since we usually have a Grapher instance in our hands and can ask it for its type. But you're right that we should find a nicer spot for this code. I'll move it, but I'm not sure where to – I'll think about it.